market_maker_keeper/ddex_market_maker_keeper.py

```diff
@@ -0,0 +1,240 @@
+
+        parser.add_argument("--debug", dest='debug', action='store_true',
+                            help="Enable debug output")
+
+        self.arguments = parser.parse_args(args)
+        setup_logging(self.arguments)
+
+        self.web3 = kwargs['web3'] if 'web3' in kwargs else Web3(HTTPProvider(endpoint_uri=f"http://{self.arguments.rpc_host}:{self.arguments.rpc_port}",
+                                                                              request_kwargs={"timeout": self.arguments.rpc_timeout}))
+        self.web3.eth.defaultAccount = self.arguments.eth_from
+        self.our_address = Address(self.arguments.eth_from)
+
+        self.pair = self.arguments.pair.upper()
+        self.token_buy = ERC20Token(web3=self.web3, address=Address(self.arguments.buy_token_address))
+        self.token_sell = ERC20Token(web3=self.web3, address=Address(self.arguments.sell_token_address))
+        self.min_eth_balance = Wad.from_number(self.arguments.min_eth_balance)
+        self.bands_config = ReloadableConfig(self.arguments.config)
+        self.price_max_decimals = None
+        self.amount_max_decimals = None
+        self.gas_price = GasPriceFactory().create_gas_price(self.arguments)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
+        self.spread_feed = create_spread_feed(self.arguments)
+        self.order_history_reporter = create_order_history_reporter(self.arguments)
+
+        self.history = History()
+        self.zrx_exchange = ZrxExchange(web3=self.web3, address=Address(self.arguments.exchange_address))
+        self.ddex_api = DdexApi(self.web3,
+                                self.arguments.ddex_api_server,
+                                self.arguments.ddex_api_timeout)
+
+        self.order_book_manager = OrderBookManager(refresh_frequency=self.arguments.refresh_frequency, max_workers=1)
+        self.order_book_manager.get_orders_with(lambda: self.ddex_api.get_orders(self.pair))
+        self.order_book_manager.cancel_orders_with(lambda order: self.ddex_api.cancel_order(order.order_id))
+        self.order_book_manager.enable_history_reporting(self.order_history_reporter, self.our_buy_orders, self.our_sell_orders)
+        self.order_book_manager.start()
+
+    def main(self):
+        with Lifecycle(self.web3) as lifecycle:
+            lifecycle.initial_delay(10)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(1, self.synchronize_orders)
+            lifecycle.on_shutdown(self.shutdown)
+
+    def startup(self):
+        self.approve()
+
+        # Get maximum number of decimals for prices and amounts.
+        # Ddex API enforces it.
+        markets = self.ddex_api.get_markets()['data']['markets']
+        market = next(filter(lambda item: item['id'] == self.pair, markets))
+
+        self.price_max_decimals = market['pricePrecision']
+        self.amount_max_decimals = market['amountDecimals']
+        #TODO minOrderSize ?
```

Comment:
We had this discussion already in the past, as some other exchanges also have the concept of minimum order size. Our decision was to not implement that in the keeper itself but rely on the fact that the `dustCutoff` in the config file will be always set to a value greater than minimum order size. So we basically assume the configuration will be done properly.

Regarding `pricePrecision` and `amountDecimal` of course they have to be here!